metadata/src/main/java/org/apache/kafka/controller/PartitionChangeBuilder.java

```diff
@@ -361,11 +410,12 @@ public Optional<ApiMessageAndVersion> build() {
 
         maybeUpdateRecordElr(record);
 
-        if (record.isr() == null && !targetIsr.isEmpty() && !targetIsr.equals(Replicas.toList(partition.isr))) {
+        if (record.isr() == null && (!targetIsr.isEmpty() || eligibleLeaderReplicasEnabled) && !targetIsr.equals(Replicas.toList(partition.isr))) {
```

Comment:
Should we add a comment here that we let ISR shrink to empty if we're using eligible leader replicas?

Reply:
Updated the comments.